public/pages/Rules/store/RulesStore.ts

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { RuleService } from '../../../services';
+import { ruleTypes } from '../utils/constants';
+import { load, safeDump } from 'js-yaml';
+import { RuleItemInfoBase, IRulesStore, IRulesCache } from '../../../../types';
+import { Rule } from '../../../../models/interfaces';
+import { NotificationsStart } from 'opensearch-dashboards/public';
+import { errorNotificationToast } from '../../../utils/helpers';
+
+/**
+ * Class is used to make rule's API calls and cache the rules.
+ * If there is a cache data requests are skipped and result is returned from the cache.
+ * If cache is invalidated then the request is made to get a new set of data.
+ *
+ * @class RulesStore
+ * @implements IRulesStore
+ * @param {BrowserServices} services Uses services to make API requests
+ */
+export class RulesStore implements IRulesStore {
+  /**
+   * Rule service instance
+   *
+   * @property {RuleService} service
+   * @readonly
+   */
+  readonly service: RuleService;
+
+  /**
+   * Notifications
+   * @property {NotificationsStart}
+   * @readonly
+   */
+  readonly notifications: NotificationsStart;
+
+  constructor(service: RuleService, notifications: NotificationsStart) {
+    this.service = service;
+    this.notifications = notifications;
+  }
+
+  /**
+   * Keeps rule's data cached
+   *
+   * @property {IRulesCache} cache
+   */
+  cache: IRulesCache = {};
+
+  /**
+   * Invalidates all rules data
+   */
+  invalidateCache = () => {
+    this.cache = {};
+    return this;
+  };
+
+  /**
+   * Returns all rules, custom and pre-packaged
+   *
+   * @method getAllRules
+   * @param {{ [key: string]: string[] }} terms
+   * @param {any} query
+   * @param {boolean} invalidateCache
+   * @returns {Promise<RuleItemInfoBase[]>}
+   */
+  public async getAllRules(
+    terms?: { [key: string]: string[] },
+    query?: any,
+    invalidateCache: boolean = false
+  ): Promise<RuleItemInfoBase[]> {
+    let prePackagedRules = await this.getRules(true, terms, query, invalidateCache);
+    let customRules = await this.getRules(false, terms, query, invalidateCache);
+
+    prePackagedRules = this.validateAndAddDetection(prePackagedRules);
+    customRules = this.validateAndAddDetection(customRules);
+
+    return customRules.concat(prePackagedRules);
+  }
+
+  /**
+   * Makes the request to get pre-packaged or custom rules
+   * Pass `invalidateCache` to invalidate cache
+   *
+   * @param {boolean} prePackaged
+   * @param {{ [key: string]: string[] }} terms
+   * @param {any} query
+   * @param {boolean} invalidateCache
+   * @returns {Promise<RuleItemInfoBase[]>}
+   */
+  async getRules(
+    prePackaged: boolean,
+    terms?: { [key: string]: string[] },
+    query?: any,
+    invalidateCache: boolean = false
```

Comment:
invalidation should happen only on select actions like create/update rule. Get should always return the cached value and does not need to provide an option for invalidation

Reply:
Yup, set only for create/update/delete.